Kubernetes (#4)
dspringuel-va/blog-posts: kubernetes/README.md

# Kubernetes in Action

Recently, I took some professional development days.

> Since Kubernetes is a black box to me, I decided to dig into it.

Reviewer: ```suggestion
Since Kubernetes was a black box to me, I decided to dig into it.
```

* [Building Blocks](#building-blocks)
  * [Container](#container)
  * [Pod](#pod)
  * [Node](#node)
  * [Cluster](#cluster)
* [Resources](#resources)
  * [Namespace](#namespace)
  * [Workloads](#workloads)
    * [Pod](#pod-1)
    * [ReplicaSet](#replicaset)
    * [Deployment](#deployment)
    * [Others](#others)
  * [Services](#services)
    * [Service (ClusterIP, NodePort, and Load Balancer)](#service)
    * [Ingress](#ingress)
  * [Config](#config)
    * [ConfigMap](#configmap)
    * [Secret](#secret)
* [Architecture](#architecture)
  * [Control Plane (a.k.a. master node)](#control-plane-aka-master-node)
    * [API Server](#api-server)
    * [etcd](#etcd)
    * [Scheduler/ Controller Manager](#scheduler--controller-manager)
  * [Worker Nodes](#worker-nodes)
    * [Kubelet / Container Runtime](#kubelet--container-runtime)
    * [Kubernetes Service Proxy (a.k.a. kube-proxy)](#kubernetes-service-proxy-aka-kube-proxy)
* [References](#references)

## What is it?

Moving from monolithic applications to a microservices architecture has many benefits: independent development and releases, granular scability, possibility to select the right technology for a problem, etc.

> However, those benefits comes at a price: the complexity of successfully managing all the services to have a performant and resiliant system.

Reviewer 2: ```suggestion
However, those benefits come at a price: the complexity of successfully managing all the services to have a performant and resiliant system.
```

Kubernetes is a platform that alleviates that complexity to make a microservice architecture easier to manage.

> From Kubernetes official website,

Reviewer 2: ```suggestion
From Kubernetes' official website,
```

Kubernetes relies on Linux containers to run heterogenous applications without having to know anything about the application internals.
This abstraction makes is easy to run multiple different applications on the same hardware.

Furthermore, the applications don't know anything about the underlying infrastructure. It is all abstracted away by Kubernetes. Thus, this simplifies development, deployment and management for the software teams.

## Building Blocks

Let's start by describing the logical building blocks of what compose a Kubernetes ensemble.

It starts with some containers, and goes all the way up to a cluster.

![Building Blocks](https://docs.google.com/drawings/d/e/2PACX-1vQXcxnzV0f-g0e9Q_XtuwdHYFFg7wmtsyQO_VsZBWyyCeJwXUY_3M5cBTjO85UWL9S0IfUuok-SDx5a/pub?w=643&h=633)

### Container

A container is a unit of software packaged with all the needed dependencies to run it.

![Container Diagram](https://docs.google.com/drawings/d/e/2PACX-1vSzE6SM-GPQvUcY6HZJ2K9kgRb9ndPv8a63k77VFgz0jlDXMKF_WYwg7Tyf9xcWJNxwUrUj_Gd19qEA/pub?w=263&h=311)

> The most common container runtime is Docker, which is one of the runtime supported by Kubernetes. rkt is an alternative to Docker, which is also supported by Kubernetes.

Reviewer 2: ```suggestion
The most common container runtime is Docker, which is one of the runtimes supported by Kubernetes. rkt is an alternative to Docker, which is also supported by Kubernetes.
```

> Docker gives the possibility to the user to build declaratively an image of the container.

Reviewer 2: ```suggestion
Docker gives the user the possibility to declaratively build an image of the container.
```

A Docker image is a model for a container.
It is then possible to create as many container as one wants based on that image.
Furthermore, it is possible to host an image on registeries over the Internet,
which thus makes container creation really easy.

> The most known one is DockerHub. Google Cloud Platform offers also its own image registry: Google Cloud Registry.

Reviewer 2: ```suggestion
The most known one is DockerHub. Google Cloud Platform also offers its own image registry: Google Cloud Registry.
```

### Pod

A Kubernetes pod is a group of 1 or more containers.

![Pod Diagram](https://docs.google.com/drawings/d/e/2PACX-1vQsoPj5L9ab2o0ZgeNqLzYQhdSiu_KAs1dNaPhwopQjujnW1SNBhX5gvyFQ2iV4_20_nqdmflNolblp/pub?w=415&h=625)

> Containers within a pod have their processes isolated (like every other containers).

Reviewer 2: ```suggestion
Containers within a pod have their processes isolated (like every other container).
```

However, containers in a pod share their network namespace, i.e. they share the same IP address and the same port space.
The IP address is chosen by Kubernetes at creation time, and only stays around for the pod instance life.

> A pod can contain more than 1 container. However, those should be tightly coupled together. Usually, that means a "main" container which run the application the pod is meant for, along with sidecar containers that provides specific value to the main container.

Reviewer: ```suggestion
A pod can contain more than 1 container. However, those should be tightly coupled together. Usually, that means a "main" container which runs the application that the pod is meant for, along with sidecar containers that provide specific values to the main container.
```

Here, main is only a logical concept.

An example could be a nginx server that runs as the main container,
along with another container that is meant to fetch the website files from somewhere else.
Here, one container can't do much without the other.

A counter-example would be to have a frontend and backend container within the same pod.
In this case, one can definitely run without the other.
Those containers shouldn't really be coupled at that level. Maybe they don't need the same resources, or same availability.
It's impossible to scale one without scaling the other, which might be a problem.
In this case, it would be better separate the apps into two pods.

### Node

A node is a machine representation where a group of pods are running.

![Node Diagram](https://docs.google.com/drawings/d/e/2PACX-1vQUI6h9QuTXPSfnJ0ehCORuNo6aknViuqi2Oan2L4a7gEajvKhv8L89qvnN8XJFSa6L-oZ7oPnjKZM5/pub?w=1517&h=851)

A pod can't live across two different nodes, i.e. have one container run in one node, and a second container in a different node.

> All pod's containers are guaranteed to run in the same node.

Reviewer 2: ```suggestion
All of a pod's containers are guaranteed to run in the same node.
```

### Cluster

A cluster is simply a group of nodes.

![Cluster Diagram](https://docs.google.com/drawings/d/e/2PACX-1vRu7wUcEbgvU8kSli7Ye58hbXlqCUKsgXVnlk-IxbSQWUpBl3tVJOCWZFGc_VQp_t1_5e3lfu5BAOWM/pub?w=1770&h=1129)

The nodes can have differents computational resources, depending on what the system needs. They also can be in different data-centers, accross different regions.

## Resources

Almost everything that defines a Kubernetes clusters is represented by a resource.
The resources are managed by a single RESTful API server.
They represent the state of the cluster at any moment in time.

> There are controllers that listens to resources CRUD operations at any time and execute actions accordingly to unify the actual state with the desired state.

Reviewer: ```suggestion
There are controllers that listen to a resource's CRUD operations at any time and execute actions accordingly to unify the actual state with the desired state.
```

That architecture makes Kubernetes declarative (instead of imperative), i.e. it only needs to be told what state the cluster should be in (and Kubernetes will take care of reconciliating it), instead of being told what to do in order to be in that state.

There are more details on Kubernetes architecture [further down](#architecture).

One last thing before jumping into individual resources, `kubectl` is a nice little CLI tool that can quickly expose documentation for each resources.
The command for it is `explain`.

For example,

```
kubectl explain pod
```

shows the documentation for the pod resource and its fields.

It is also possible to dig deeper in the doc.
If one wants to know more about pod spec, the command to run is the following:

```
kubectl explain pod.spec
```

### Namespace

> Every single resource in the cluster is named. The namespace resource is used to group resources togheter in a non overlapping way. Therefore, all resources' name must be unique within a namespace. A different resource with the same name can exist in a different namespace.

Reviewer 2: ```suggestion
Every single resource in the cluster is named. The namespace resource is used to group resources together in a non overlapping way. Therefore, all resources' names must be unique within a namespace. A different resource with the same name can exist in a different namespace.
```

![Namespace Diagram](https://docs.google.com/drawings/d/e/2PACX-1vTR7MCsBMEVbMBA9AQg-32kEKqNLHMdkK1sunii7x2Rb5fylYdhwYszPjYGEP2QqxReeIwnzGloRcVk/pub?w=1898&h=1014)

Kubernetes creates a default namespace, where all resources go if no namespace is specified upon resource creation.

Some resources are cluster-level resources, which means that they can't be in a namespace.

> The namespace resource itself is a good example of a cluster-level resource (Kubernetes doesn't allow to have a namespace resource within a namespace).

Reviewer 2: ```suggestion
The namespace resource itself is a good example of a cluster-level resource (Kubernetes doesn't you allow to have a namespace resource within a namespace).
```

Author: Did you mean `... doesn't allow you to ...`?

Reviewer 2: Ha, yeah.

Nodes and persistent volumes are other examples of cluster-level resources.

### Workloads

> Workloads is a logical group of resources that manages the life of pods. The most simple one is the pod resource itself. Let's take a look at some few of those resources.

Reviewer 2: ```suggestion
Workloads is a logical group of resources that manages the life of pods. The most simple one is the pod resource itself. Let's take a look at a few of those resources.
```

#### Pod

The Pod is the atomic workload resource. As mentioned before, it runs one or more containers within it.

![Detailed Pod Diagram](https://docs.google.com/drawings/d/e/2PACX-1vRO_zH_xjWw9SXZ9szPmCUtxYmg52Sj8TscNJ8dBBiAids0XFRFxxVu1NfRwEuGQNTph79P3cOuTpBy/pub?w=824&h=870)

In the Pod resource definition, it is possible to attribute many properties.
The metadata exposes information on the pod itself, like its name, namespace or labels.
The spec describes among other things which container to run, from which images to create them, or which volume to create.

> Labels are part of the metadata section of the pod definition, and it is a important property for the pods.

Reviewer 2: ```suggestion
Labels are part of the metadata section of the pod definition, and it is an important property for the pods.
```

It allows the user to organize the pods, in a potentially overlapping way.
Furthermore, it allows other resources to select pods based on those labels.

The ReplicaSet is one of such resource.

#### ReplicaSet

> A ReplicaSet is a resource responsible to manage a group of pods. The main properties of that resources are the pod template (what each pod should be like), the pod selector (which pods are part of this set), and the replicas (how many pods should be running under that set).

Reviewer 2: ```suggestion
A ReplicaSet is a resource responsible for managing a group of pods. The main properties of that resources are the pod template (what each pod should be like), the pod selector (which pods are part of this set), and the replicas (how many pods should be running under that set).
```

![Replica Set Diagram](https://docs.google.com/drawings/d/e/2PACX-1vTCrgcAKk-InhOQPw84ULsK8Z0eioosmABeYZYDZzKa8gyY9vhwUr1mk0FYscNq7cAvxOHgyfSqyFS3/pub?w=991&h=1064)

> The ReplicaSet watches for any pod changes. If for any reasons, the number of pods doesn't match the replica number (a pod crashed, a pod label changed, or the host node failed for example), the ReplicaSet will spin up a new pod automatically.

Reviewer: ```suggestion
The ReplicaSet watches for any pod changes. If, for any reason, the number of pods doesn't match the replica number (a pod crashed, a pod label changed, or the host node failed for example), the ReplicaSet will spin up a new pod automatically.
```

The ReplicaSet doesn't decide on which node the pod is created, nor its IP address.
It only cares about the number of pods running.
If the ReplicaSet replicas property would change (up or down), it would then command to create or kill the needed pods.

> If the pod template would change, it doesn't affect the currently running pods under the ReplicaSet. The next the ReplicaSet needs to create a pod, it would use the new template, which means that two versions of the pods could concurrently run under that ReplicaSet (which isn't ideal).

Reviewer: ```suggestion
If the pod template would change, it doesn't affect the currently running pods under the ReplicaSet. The next time the ReplicaSet needs to create a pod, it would use the new template, which means that two versions of the pods could concurrently run under that ReplicaSet (which isn't ideal).
```

The Deployment resource alleviates that problem.

#### Deployment

The Deployment resource is almost the same as the ReplicaSet. It has a pod selector, a pod template and a replicas property.

![Deployment Diagram](https://docs.google.com/drawings/d/e/2PACX-1vRUwVm0eMdmQVOkQ6WOR6kiMRP_lM1cBQgFixzWFAh-l5Ml2jPKRVgYXxx2tXzY-pFap4Q5_1ai6YzV/pub?w=4487&h=1709)

> The difference resides on what happens when the pod template is updated.

Reviewer 2: ```suggestion
The difference resides in what happens when the pod template is updated.
```

In that scenario, the Deployment will create another ReplicaSet with the updated pod template.
That new ReplicaSet has its replicas property set to 0.

> Then, it gradually increase the new ReplicaSet replicas number, while decreasing the old ReplicaSet replicas number to 0.

Reviewer 2: ```suggestion
Then, it gradually increases the new ReplicaSet replicas number, while decreasing the old ReplicaSet replicas number to 0.
```

![Deployment Diagram - Rolling Update](https://docs.google.com/drawings/d/e/2PACX-1vTKew2mhzCXOMfBZOHCxyF3eOykFSL9h-PKkbj2RmrkkdlpOR7U4IiFdzSPiy3OwMUeQvVdKErcmm-i/pub?w=3325&h=1699)

Once the rolling update is done, the Deployment doesn't delete the old ReplicaSet. It is kept around, in case that a rollback is needed.

> The Deployment defines also a Rolling Update Strategy. It is compose mainly of two properties: maxSurge and maxUnavailable.

Reviewer 2: ```suggestion
The Deployment defines also a Rolling Update Strategy. It is composed mainly of two properties: maxSurge and maxUnavailable.
```

The maxSurge property defines how many more pods than the desired number can run during the update.
The maxUnavailable property defines the opposite, i.e. how many fewer pods than the desired number can run during the update.

For example, if the maxSurge is set to 1 and maxUnavailable is set to 0 with a desired number set to 3, the rolling update will have at most 4 pods at any moment, and at least 3 at any moment.

#### Others

There are many other resources that can spin up new pods. Let's talk about a few of them.

*Job*

> A Job is almost the same resource than a Pod.

Reviewer 2: ```suggestion
A Job is almost the same resource as a Pod.
```

Usually, a Pod needs to run continuously. When it fails, the general use case for it is to be brought back to life, usually by a ReplicaSet.

> However, there are some case where a pod would normally execute and terminate successfully.

Reviewer 2: ```suggestion
However, there are some cases where a pod would normally execute and terminate successfully.
```

> In those case, the pod shouldn't be restarted, since it was meant for a one time execution.

Reviewer 2: ```suggestion
In those cases, the pod shouldn't be restarted, since it was meant for a one time execution.
```

The Job resource was created especially for that purpose.

The Job will run a pod until it finishes successfully. If the pod fails for any reason (the node crashed, etc), it will reschedule it to run again.

Furthermore, it is possible to setup a job to run many times the pods, sequentially or concurrently (or both).

*CronJob*

> A CronJob is the exact same thing than a Job, except the fact that is possible to schedule the running time of the pod in time.

Reviewer 2: ```suggestion
A CronJob is the exact same thing as a Job, except for the fact that it is possible to schedule the running time of the pod in time.
```

Furthermore, it makes sure that the service load is evenly distributed amongst all nodes.

#### Ingress

When there are many different services exposed externally, it can become hard to manage all the load-balancer's IP addresses.

An Ingress is a resources that sits in front of many other services.
It operates at the L7 (application layer) of the network stack.
That means that it can redirect requests based on either the request domain, or the request path for example.

![Ingress Diagram](https://docs.google.com/drawings/d/e/2PACX-1vQ5gIevkt6uJntEbEBWg0sIMOD8m6dBw6EJ-x0X4jdvRJuvK0b-LegHQY2VofHJrjObNKRUCdvQn4X_/pub?w=2003&h=823)

To make the Ingress resources work, it needs a Ingress controller. For example, GKE uses Google Cloud Platform's own HTTP load-balancing features to provide the Ingress functionality.

With the correct set of rules, it would be possible to have a single Ingress resource that act as an entry point to all the services within the cluster.

### Config

Almost every application needs some configuration.
Usually, we don't want that configuration hard-coded in the application itself.

> Kubernetes defines two resources to help passing configurations to application: ConfigMaps and Secrets.

Reviewer 2: ```suggestion
Kubernetes defines two resources to help passing configurations to applications: ConfigMaps and Secrets.
```

#### ConfigMap
A ConfigMap resource is simply a list of key-value pairs.

> It can be passed in the pod either by environments variable or by a mounted volume.

Reviewer 2: ```suggestion
It can be passed in the pod either by environment variable or by a mounted volume.
```

## Architecture

Now that a we know a bit more about what Kubernetes can do, let's take a look on how it works under the hood.

![Overall Architecture Diagram](https://docs.google.com/drawings/d/e/2PACX-1vQCYKzEwsN1q74BkfHv85o6bsE_1gSigon8mHLn9Yf7A4bIYEV0PRT0Y5GT97_Sss-i51NL9LwJysC0/pub?w=1416&h=778)

Kubernetes' architecture consists of a Control Plane and worker nodes (or plainly, the nodes). Some system applications are also running on the worker nodes.

### Control Plane (a.k.a. Master Node)

The Control Plane consists of an API service, an etcd database, a scheduler and a controller manager. Let's take a more detailed look at each of those elements.

#### API Server

The API Server is the focal point of the Control Plane.
It is simply a RESTful API server that takes care of managing all the Kubernetes resources (basic CRUD operations).
It also offers a "watch" functionality, so take clients can listen on resources updates they care about.

#### etcd

Resources needs to be persistently stored somewhere.

> That where etcd comes in.

Reviewer 2: ```suggestion
That is where etcd comes in.
```

The API Server stores all the resources directly in etcd, and is the only part of the system that has access to it.

#### Scheduler / Controller Manager

Having a simple CRUD server without doing anything with that data wouldn't do much work.
Various controllers exist to run reconciliation tasks to make sure the desired state reflects the actual cluster state.

Those controllers are run in the Controller Manager.
For example, when the ReplicaSet controller picks up new/updated ReplicaSet, it create the necessary Pod resources accordingly.
That specific controller doesn't create the actual Pod, only the Pod resources.
Creating the actual pods is made by another component in the system, in response of a new/updated Pod resources.

The Scheduler is a specific type of controller that takes care of choosing on which node each pod should run.
It listens on Pod resource creation.
When it happens, the Scheduler finds all the nodes that can accept the Pod (based on resources management, or affinity for example), prioritizes those nodes and choose the best one.
Then, it updates the Pod resource to indicates which node it should run on.

> There are almost a controller for every type of resource creatable. Furthermore, every controller's job is really scoped to the resource it's controlling. It makes the controller loosely coupled from one another.

Reviewer 2: ```suggestion
There are almost a controller for every type of resource creatable. Furthermore, every controller's job is really scoped to the resource it's controlling. It makes the controller loosely coupled from one another.
```
```suggestion
There is almost a controller for every type of resource creatable. Furthermore, every controller's job is really scoped to the resource it's controlling. It makes the controller loosely coupled from one another.
```

### Worker Nodes

> The API server and the controllers is at the heart of the Control Plane. However, some system components must run also on each of the worker nodes, namely the Kubernetes Service Proxy and the Kubelet.

Reviewer 2: ```suggestion
The API server and the controllers are at the heart of the Control Plane. However, some system components must also run on each of the worker nodes, namely the Kubernetes Service Proxy and the Kubelet.
```

#### Kubelet / Container Runtime

The Kubelet is responsible for everything running on the worker nodes, but
the main responsibility is to manage the life of containers on the node.

When a pod is scheduled on a node, Kubelet gets notified and actually starts the container by using the configured container runtime (Docker, etc).

> Also, Kubelet is responsible for monitoring the health of the pods, either by running liveness and readiness probes, or by watching computational resources consumption.

Reviewer 2: ```suggestion
Also, Kubelet is responsible for monitoring the health of the pods, either by running liveness and readiness probes, or by watching computational resource consumption.
```

It also stops the container when the corresponding pod resource gets deleted from the API server.

> Finally, the Kubelet is also responsible with registering the node to the cluster by creating a Node resource on the API server.

Reviewer 2: ```suggestion
Finally, the Kubelet is also responsible for registering the node to the cluster by creating a Node resource on the API server.
```
